File: src/main/java/com/tambapps/math/fourier/filtering/ThresholdFilter.java
package com.tambapps.math.fourier.filtering;

import com.tambapps.math.complex.Complex;

/**
 * A threshold filter. Filters values that are greater/lower than a threshold
 */
class ThresholdFilter extends AbstractFilter {

  private final double threshold;
  private final boolean filterLower;

  /**
   * Creates a threshold filter
   *
   * @param threshold   the threshold
   * @param filterLower whether to filter values lower or greater than the threshold
   */
  ThresholdFilter(double threshold, boolean filterLower) {
    this.threshold = threshold;
    this.filterLower = filterLower;
  }

  @Override
  Complex apply(Complex c, int i, int j, int M, int N) {
    if (filterLower) {
      return c.abs() < threshold ? c : Complex.ZERO;
    } else {
      return c.abs() > threshold ? c : Complex.ZERO;
    }
  }
}

File: src/main/kotlin/com/tambapps/fft4j/fourier/filtering/ThresholdFilter.kt
package com.tambapps.fft4j.fourier.filtering

import com.tambapps.fft4j.complex.Complex

/**
 * A threshold filter. Filters values that are greater/lower than a threshold
 */
class ThresholdFilter
/**
 * Creates a threshold filter
 *
 * @param threshold   the threshold
 * @param filterLower whether to filter values lower or greater than the threshold
 */(private val threshold: Double, private val filterLower: Boolean) : AbstractFilter() {

    override fun apply(c: Complex, i: Int, j: Int, M: Int, N: Int): Complex {
        return if (filterLower) {
            if (c.abs() < threshold) c else Complex.ZERO
        } else {
            if (c.abs() > threshold) c else Complex.ZERO
        }
    }

}